utils/pb.go

```
@@ -0,0 +1,182 @@
+
+		tuple.writeMetric.Entries = append(tuple.writeMetric.Entries, writeEntry)
+	}
+
+	writeRequest := &ceresdbproto.WriteRequest{
+		Metrics: make([]*ceresdbproto.WriteMetric, 0, len(tuples)),
+	}
+	for _, tuple := range tuples {
+		tuple.writeMetric.TagNames = tuple.tagDict.toOrdered()
+		tuple.writeMetric.FieldNames = tuple.fieldDict.toOrdered()
+		writeRequest.Metrics = append(writeRequest.Metrics, &tuple.writeMetric)
+	}
+	return writeRequest
+}
+
+func buildPbValue(v interface{}) *ceresdbproto.Value {
+	switch v.(type) {
+	case bool:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_BoolValue{
+				v.(bool),
+			},
+		}
+	case string:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_StringValue{
+				v.(string),
+			},
+		}
+	case float64:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Float64Value{
+				v.(float64),
+			},
+		}
+	case float32:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Float32Value{
+				v.(float32),
+			},
+		}
+	case int64:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Int64Value{
+				v.(int64),
+			},
+		}
+	case int32:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Int32Value{
+				v.(int32),
+			},
+		}
+	case int16:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Int16Value{
+				int32(v.(int16)),
+			},
+		}
+	case int8:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Int8Value{
+				int32(v.(int8)),
+			},
+		}
+	case uint64:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Uint64Value{
+				v.(uint64),
+			},
+		}
+	case uint32:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Uint32Value{
+				v.(uint32),
+			},
+		}
+	case uint16:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Uint16Value{
+				uint32(v.(uint16)),
+			},
+		}
+	case uint8:
+		return &ceresdbproto.Value{
+			Value: &ceresdbproto.Value_Uint8Value{
+				uint32(v.(uint8)),
+			},
+		}
+	default:
+		return nil
```

return err here?

---

precheck is done in build, default branch will never be entered

---

If it's already prechecked, I think we can just `panic("not reachable")` instead of return a `nil` value, which is invalid itself.

---

ok，add err in return val